Document how to use uenvs as upstream Spack instances

RMeli opened the pull request:
Draft for a documentation page explaining how to use user environments as upstream Spack instances. I think many people at CSCS use this or a similar workflow for development, so it might be useful for the end users too.

Feedback and suggestions on how to improve it very welcome! This is a rough initial draft to get the conversation started.

simonpintarelli commented on docs/uenv-compilation-spack.md:
```diff
@@ -0,0 +1,163 @@
+
+### Spack Environment and Building Dependencies
+
+As described above, you can define a [Spack environment] describing the version of the package you want to build and the contraint on the dependencies. After concretizing the environment with
```
typo: `constraint`

RMeli replied:
```suggestion
As described above, you can define a [Spack environment] describing the version of the package you want to build and the constraint on the dependencies. After concretizing the environment with
```

simonpintarelli commented:
Why not using a develop spec (`spack develop -p ...`) in an anonymous environment instead of `spack <spec> build-env -- bash` and then manually calling the build system?

The only disadvantage I can see is that with the develop spec one has to edit the recipe if the API of the build system has changed.

simonpintarelli commented on docs/uenv-compilation-spack.md:
````diff
@@ -0,0 +1,163 @@
+cp2k@2024.1 +cuda cuda_arch=80 +cosma ^cosma +gpu_direct
+```
+
+COSMA is not availabe in the uenv described above since it is not a dependency needed by `cp2k@2024.1 +cuda cuda_arch=80`. 
````
It might be worth mentioning here that in general it is `cuda_arch=90`, otherwise people might copy/paste the wrong arch.

RMeli replied:
Yeah, I realized too late this is an annoying one... `cp2k@2024.1` can't be build with `cuda_arch=90`. Maybe worth using a different example (like SIRIUS) where we can actually use `cuda_arch=90`?

simonpintarelli replied:
Yes, or QuantumESPRESSO%nvhpc.

Or adding `packages:all:prefer:cuda_arch=90`, but then set it explicitly to 80 in the spec and explain why.

simonpintarelli replied:
For QE I've added the spack upstream example in the PR: https://github.com/eth-cscs/alps-uenv/pull/92
But it's not as well explained as you did.

RMeli replied:
Thanks for pointing it out, I missed it. Having it in a central location would avoid repetition, since it's useful for many uenvs. I'll look at what you wrote already, and incorporate it with this PR. I'll switch package too, to avoid confusion.

> Or adding packages:all:prefer:cuda_arch=90, but then set it explicitly to 80 in the spec and explain why.

This is what I did for the recipe, but is an additional layer of information that a general user does not need to know about (it's CP2K-specific).

simonpintarelli commented on docs/uenv-compilation-spack.md:
```diff
@@ -0,0 +1,163 @@
+
+### Spack Environment
+
+To make things clean and reproducible, we use [Spack Environments] do describe what we want to build. To define a [Spack environment] we create the following file in a folder (hereafter referred to as `SPACK_ENV_FOLDER`):
```
It is missing that the file must be named `spack.yaml`.

What about giving the commands `spack env create`, `spack  -e [..] add <spec>`
`spack config add [..]` to create `spack.yaml`?

msimberg replied:
In addition to `spack env create <name>`, you can also do `spack -e <name> config edit` to dump the below contents into the file. Not sure which is simpler to grok for someone not used to envs though.

simonpintarelli commented on docs/uenv-compilation-spack.md:
```diff
@@ -0,0 +1,163 @@
+    This guide assumes that you have a local installation of [Spack]. If you don't have [Spack] installed, follow [Spack Getting Started].
+
+!!! warning
+    Avoid installing [Spack] on `HOME`. Packages are installaed within the `spack/` folder, and you might quickly run out of space.
```
A note about compatible spack versions might be useful. Most of our current envs have been built with spack v0.21.2, which won't work as upstream if used with spack@develop.

RMeli replied:
Great point!

msimberg commented on docs/uenv-compilation-spack.md:
```diff
@@ -0,0 +1,163 @@
+# Using uenvs as upstream Spack instances
+
+User-environments (uenvs) are built with [Spack] using the [Stackinator] tool. Therefore, an `uenv` is tightly copuled with [Spack] and can be used as an upstream [Spack] instance (see [Chaining Spack Installations] for more details).
```
```suggestion
User-environments (uenvs) are built with [Spack] using the [Stackinator] tool. Therefore, a `uenv` is tightly coupled with [Spack] and can be used as an upstream [Spack] instance (see [Chaining Spack Installations] for more details).
```

General question which doesn't need to be answered here: Is there a distinction between
- user-environment
- uenv
- `uenv`

?

Naively I'd say the first is a generic term, same as user environment. uenv: project name or a squashfs image? `uenv`: name of the command line tool?

msimberg commented on docs/uenv-compilation-spack.md:
```diff
@@ -0,0 +1,163 @@
+    This guide assumes that you have a local installation of [Spack]. If you don't have [Spack] installed, follow [Spack Getting Started].
+
+!!! warning
+    Avoid installing [Spack] on `HOME`. Packages are installaed within the `spack/` folder, and you might quickly run out of space.
```
```suggestion
    Avoid installing [Spack] on `HOME`. Packages are installed within the `spack/` folder, and you might quickly run out of space.
```

msimberg commented on docs/uenv-compilation-spack.md:
```diff
@@ -0,0 +1,163 @@
+    This guide assumes that you have a local installation of [Spack]. If you don't have [Spack] installed, follow [Spack Getting Started].
+
+!!! warning
+    Avoid installing [Spack] on `HOME`. Packages are installaed within the `spack/` folder, and you might quickly run out of space.
```
The eternal question: do we already know where spack should be installed by users, instead of saying "don't use home"? This is not a blocking comment. We might have to live with this until we know better.

RMeli replied:
@simonpintarelli suggested `$SCRATCH` in #92. Should we go for that for the time being, and then update the documentation when there is a final decision?

msimberg commented on docs/uenv-compilation-spack.md:
````diff
@@ -0,0 +1,163 @@
+cp2k@2024.1 +cuda cuda_arch=80 +cosma ^cosma +gpu_direct
+```
+
+COSMA is not availabe in the uenv described above since it is not a dependency needed by `cp2k@2024.1 +cuda cuda_arch=80`. 
````
```suggestion
COSMA is not available in the uenv described above since it is not a dependency needed by `cp2k@2024.1 +cuda cuda_arch=80`. 
```

msimberg commented on docs/uenv-compilation-spack.md:
````diff
@@ -0,0 +1,163 @@
+spack -e SPACK_ENV_FOLDER concretize -f
+```
+
+The result of the concretization will be printed on screen. Packages starting with `[+]` are packages that will be freshly installed in your local [Spack] instance. Packages marked ad `[e]` (external) are packages taken directly from the uenv (which we are using as upstream [Spack] instance). You should see many packages marked as `[e]`, which are being re-used from the uenv. This will greatly sp ...
````
```suggestion
The result of the concretization will be printed on screen. Packages starting with `[+]` are packages that will be freshly installed in your local [Spack] instance. Packages marked as `[e]` (external) are packages taken directly from the uenv (which we are using as upstream [Spack] instance). You should see many packages marked as `[e]`, which are being re-used from the uenv. This will greatly speed up compilation, since [Spack] will have to build only a small subset of packages.
```

msimberg commented on docs/uenv-compilation-spack.md:
````diff
@@ -0,0 +1,163 @@
+spack -e SPACK_ENV_FOLDER concretize -f
+```
+
+The result of the concretization will be printed on screen. Packages starting with `[+]` are packages that will be freshly installed in your local [Spack] instance. Packages marked ad `[e]` (external) are packages taken directly from the uenv (which we are using as upstream [Spack] instance). You should see many packages marked as `[e]`, which are being re-used from the uenv. This will greatly sp ...
````
Packages that aren't yet installed have ` - `, while packages that are not external but are already installed have `[+]`. The latter shouldn't show up for someone doing this guide for the first time so it can perhaps be left out, but I don't think it hurts to mention it either.

RMeli replied:
Yep. It is also `[^]` for upstream packages, not `[e]`, which is for externals. I'll correct and expand.

msimberg commented on docs/uenv-compilation-spack.md:
```diff
@@ -0,0 +1,163 @@
+## Known Limitations
+
+!!! warning
+    Swapping the upstream [Spack] instance by loading different uenvs might lead to a corruption of the [Spack] database. If this happens, you can uninstall everything from your local [Spack] instance with `spack uninstall --all` and clean up with `spack clean --all`.
```
```suggestion
    Swapping the upstream [Spack] instance by loading different uenvs might lead to surprising inconsistencies in the [Spack] database. If this happens, you can uninstall everything from your local [Spack] instance with `spack uninstall --all` and clean up with `spack clean --all`.
```
maybe? This warning should be scary, but "corruption" may be a bit too strong (since `spack reindex` can make it somewhat usable again with a new uenv).

simonpintarelli replied:
I'm wondering if we should add `config:install_tree:root:$spack/<hash of uenv recipe>/opt/spack` to `/user-environment/config.yaml`? Then each uenv would have it's own spack install tree. Or if it's better to let the user manage it consciously by creating different spack clones on scratch (if needed at all).

msimberg reviewed (commented):
Very nice, added some minor comments.

RMeli commented:
> Why not using a develop spec (spack develop -p ...) in an anonymous environment instead of spack <spec> build-env -- bash and then manually calling the build system?

I added instructions for both, so the end user can choose.

> The only disadvantage I can see is that with the develop spec one has to edit the recipe if the API of the build system has changed.

I added a warning explaining this.

simonpintarelli commented:
A frequent use case might also be that the user wants to build an older version, or a different configuration, e.g. QE+qmcpack and similar. Might be good to point out that the anonymous environment is the right thing for that too.

RMeli commented:
> A frequent use case might also be that the user wants to build an older version, or a different configuration, e.g. QE+qmcpack and similar.

This should be covered by the first part of the documentation (i.e. `quantum-espresso +mpigpu`)? Otherwise I'm not sure what you mean.

Probably the phrasing at the beginning should be changed, because we don't describe only a developer workflow. I can add a note that if the is a strong feature request they should open a ticket and we could also update the official environment?

RMeli commented:
Do you think it's worth to add a note/distinguish between anonymous and managed Spack environment? I think I should be more careful to be more consistent and only use one.

bcumming reviewed (approved):
This looks very good to me.
I would like to merge it, and then others can make changes through separate pull requests.

simonpintarelli commented:
> > A frequent use case might also be that the user wants to build an older version, or a different configuration, e.g. QE+qmcpack and similar.
> 
> This should be covered by the first part of the documentation (i.e. `quantum-espresso +mpigpu`)? Otherwise I'm not sure what you mean.
> 
> Probably the phrasing at the beginning should be changed, because we don't describe only a developer workflow. I can add a note that if the is a strong feature request they should open a ticket and we could also update the official environment?

Sorry for the delay I missed the email notification. Indeed the +mpigpu example is what I meant, it is already there too.
